### PR TITLE
Improve tests semantics

### PR DESCRIPTION
**Pull request by angel-daza**
This PR replaces self-referential tests with externally anchored ones, including an update to the external validation scripts run against `braincoder` and `pyprf`. While applying this changes two extra
bugs were found: 
- A 3.2% stretch in the HRF sampling grid. (See `src/prfmodel/impulse/base.py`)
- A unit-axis double flip that paired each unit with another unit's impulse response. (See `src/prfmodel/impulse/_convolve.py`)

Both bugs are fixed here as well.

More things added:
- `tests/reference/_oracle.py`: an independent NumPy/SciPy implementation of the forward
  pRF pipeline (RF → encoding → HRF → convolution), written from definitions rather than
  transcribed from the package.
- `validate-packages.yml` now runs the validation scripts every week to make sure they are still consistent with the current state of the package

Main issue addresed here (from `ISSUES.md`):

- H3. The test suite cannot detect model-semantics bugs

While fixing H3 these issues were also fixed:
- C4. BaseImpulse.frames ignores resolution
- "C5." (NEW) Each unit was convolved with another unit's HRF (double flip)
- H2. validation/ — the only external ground-truth check — is broken and never runs
- M5. CF Gaussian normalization: comment contradicts code (code was already corrected on main; this branch added the guard test)
- N3. The impulse time grid is always built at floatx() precision, ignoring dtype

### CODE REVIEW

File: tests/reference/test_convolution.py [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+                for i in range(3)
+            ],
+        )
+
+        np.testing.assert_allclose(batched, alone, rtol=1e-8)
+
+
+class TestPaddingConvention:
+    """Tests the edge-padding choice, which diverges from prfpy and braincoder."""
+
```
Comment (maltelueken) [inline tests/reference/test_convolution.py]: Both prfpy and braincoder also pad the pRF response by repeating the first frame, so this claim is not true. I am also not sure whether this test is truly necessary.

File: tests/reference/test_convolution.py [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+def _reference_convolve(response: np.ndarray, kernel: np.ndarray) -> np.ndarray:
+    """Causal convolution with edge padding, built from `numpy.convolve`.
+
+    Equivalent to `out[t] = sum_j kernel[j] * response[t - j]` with `response[i < 0]` taken to be
+    `response[0]`.
+    """
+    pad_len = kernel.size - 1
+    padded = np.pad(response, (pad_len, 0), mode="edge")
+    return np.convolve(padded, kernel)[pad_len : pad_len + response.size]
```
Comment (maltelueken) [inline tests/reference/test_convolution.py]: This function is an exact duplicate from `_oracle.convolve`. It would perhaps be better to define it in a single location (e.g. import from `_oracle` here).

File: src/prfmodel/impulse/base.py [hunk 3]
```diff
@@ -95,13 +96,47 @@ def frames(self) -> Tensor:
...
         The time frames at which the impulse response function is evaluated.
 
-        Time frames are linearly interpolated between `offset` and `duration` and have shape `(1, num_frames)`.
+        Time frames start at `offset` and are spaced `resolution` apart, with shape `(1, num_frames)`, at
+        :func:`keras.config.floatx` precision. Use :meth:`get_frames` to build them at another precision.
 
         """
-        if self._frames is None:
-            self._frames = ops.expand_dims(ops.linspace(self.offset, self.duration, self.num_frames), 0)
+        return self.get_frames()
```
Comment (maltelueken) [inline src/prfmodel/impulse/base.py]: I think this property is not needed anymore since `get_frames` allows control over the dtype.

File: .github/workflows/validate-packages.yml [hunk 4]
```diff
@@ -1,8 +1,16 @@
...
+#
+# A manual run validates the single package chosen from the dropdown. A scheduled run supplies no
+# inputs, so it falls back to validating every supported package -- see the `package` matrix below.
+# These checks compare prfmodel against independent implementations and are the only thing in the
+# repository that can catch a convention diverging from the field; the in-repository test suite
+# cannot, because it verifies prfmodel against itself.
 on:
+  schedule:
+    # Mondays at 06:00 UTC.
+    - cron: '0 6 * * 1'
```
Comment (maltelueken) [inline .github/workflows/validate-packages.yml]: Depending on how long this action takes to run, I think it might be better to trigger it with every push or PR to main (as done with the other actions).

File: tests/impulse/test_reference_hrf.py
Comment (maltelueken) [inline tests/impulse/test_reference_hrf.py:1]: I think it might be more coherent to move this file into the `tests/reference` folder.

File: src/prfmodel/impulse/base.py
Comment (maltelueken) [inline src/prfmodel/impulse/base.py]:
```suggestion
        offset: float = 0.0,
```

File: src/prfmodel/impulse/base.py [hunk 5]
```diff
@@ -95,13 +96,47 @@ def frames(self) -> Tensor:
...
+        The axis is built at the requested precision rather than cast from a cached one, so a `float64`
+        request carries `float64` time values and not widened `float32` ones. Results are cached per dtype.
+        The cache is not invalidated if `duration`, `offset` or `resolution` are reassigned after
+        construction.
+
+        """
+        dtype = get_dtype(dtype)
+
+        if dtype not in self._frames:
+            grid = ops.arange(self.num_frames, dtype=dtype) * self.resolution + self.offset
```
Comment (maltelueken) [inline src/prfmodel/impulse/base.py]:
```suggestion
            grid = ops.arange(self.num_frames, dtype=dtype) * self.resolution + self.resolution / 2 + self.offset
```

Setting the default offset to zero and adding `self.resolution / 2` means that `_frames` contains the mid-point (and not the left edge) of every frame at which the impulse response is evaluated. This leads to more accurate convoluted signals.

File: src/prfmodel/impulse/base.py [hunk 6]
```diff
@@ -95,13 +96,47 @@ def frames(self) -> Tensor:
...
+        `duration` is an upper bound: the axis holds `num_frames = int(duration / resolution)` samples spaced
+        exactly `resolution` apart, so it ends at the last whole sample at or below `duration` rather than at
+        `duration` itself. At the defaults that is 32 samples from 0 to 31 seconds. Note that nilearn differs:
+        ``spm_hrf(time_length=32)`` spans 0 to 32 seconds in 32 samples, so its spacing is `32 / 31` seconds
+        rather than the requested 1.
+
+        The axis is built at the requested precision rather than cast from a cached one, so a `float64`
+        request carries `float64` time values and not widened `float32` ones. Results are cached per dtype.
+        The cache is not invalidated if `duration`, `offset` or `resolution` are reassigned after
+        construction.
```
Comment (maltelueken) [inline src/prfmodel/impulse/base.py]: Maybe it's worth turning `duration`, `offset`, and `resolution` into read-only properties.

File: src/prfmodel/impulse/base.py [hunk 7]
```diff
@@ -95,13 +96,47 @@ def frames(self) -> Tensor:
...
+            Time frames of shape `(1, num_frames)` and dtype `dtype`, starting at `offset` and spaced
+            `resolution` apart.
+
+        Notes
+        -----
+        `duration` is an upper bound: the axis holds `num_frames = int(duration / resolution)` samples spaced
+        exactly `resolution` apart, so it ends at the last whole sample at or below `duration` rather than at
+        `duration` itself. At the defaults that is 32 samples from 0 to 31 seconds. Note that nilearn differs:
+        ``spm_hrf(time_length=32)`` spans 0 to 32 seconds in 32 samples, so its spacing is `32 / 31` seconds
+        rather than the requested 1.
```
Comment (maltelueken) [inline src/prfmodel/impulse/base.py]: I would move this to the notes of the class. I also think that the difference to nilearn is not necessary to mention here. We can mention this in a tutorial maybe.

File: src/prfmodel/impulse/_convolve.py
Comment (maltelueken) [inline src/prfmodel/impulse/_convolve.py]: This function should mention that both responses should be sampled at the same rate.

**Review by maltelueken — CHANGES_REQUESTED** [review]
Looks very good! Also great that you caught a new bug. I made a few comments. The main one is about the spacing of the frames in the impulse model.

We can have a chat about the comments if you like.

File: .github/workflows/validate-packages.yml (hunk 4 above)
Comment (angel-daza) [reply .github/workflows/validate-packages.yml]: so the "problem" with adding this to each PR is that it needs to install the prfpy and braincoder to run it each time. I think it might be annoying while we are making so many changes?

**Comment by angel-daza** [comment]
> Looks very good! Also great that you caught a new bug. I made a few comments. The main one is about the spacing of the frames in the impulse model.
> 
> We can have a chat about the comments if you like.

The workflow for external validation now runs (quite efficiently) every PR. The frames are now measured at the middle, which also created a divergence with the `prfpy` package. The validation code was adjusted to compensate for this differences, worth to check that specifically.

For the rest of changes it is mostly what the comments requested.

**Review by maltelueken — APPROVED** [review]
Looks good, thanks!
